content-src/static/DocumentChangeMessenger.js

```diff
@@ -0,0 +1,51 @@
+/* globals content addEventListener sendAsyncMessage */
+
+/**
+ * Listen for document unloads and send back the document content
+ */
+function DocumentUnloadMessenger() {
+  this.cfmm = null;
+}
+
+DocumentUnloadMessenger.prototype = {
+  /**
+   * Initializes the object.
+   *
+   * @param {nsIContentFrameMessageManager} cfmm
+   *   the content frame message manager (frame script global).
+   */
+  init(cfmm) {
+    this.cfmm = cfmm;
+    // Reattach unload listener to the window for every new page loaded.
+    this.cfmm.addEventListener("DOMContentLoaded", (this._addWindowListeners).bind(this), false);
+  },
+
+  /**
+   * Add an unload listener to the main window only.
+   *
+   * @param {Event} event
+   *   the DOMContentLoaded event object.
+   */
+  _addWindowListeners(event) {
+    let window = event.target.defaultView;
+    if (window === this.cfmm.content) { //we don't want get iframes
+      window.addEventListener("unload", this._sendContentMessage, false);
+    }
+  },
+
+  /**
+   * Send a message to the chrome process with the document content.
+   *
+   * @param {UIEvent} event
+   *   the unload event object.
+   */
+  _sendContentMessage(event) {
+    sendAsyncMessage("loci@document-unloading", {
```

Comment:
might want to have a different message structure. For instance, we could have the message name to be something like `loci-document-changes`, and the payload to contain the message type, e.g.:

```javascript
{
  type: "document-content",
  data: {
    url: "...",
    data: "..."
  }
}
```

The benefit is that we'll only need to setup one listener downstream for any event that we collect.

Note that with my naming scheme I try to only encode the "what" information as opposed to the "how".
It is not relevant for a listener to know that we collect the data upon unloading, for instance.